src/qibolab/transpilers/routing.py

```diff
@@ -0,0 +1,339 @@
+        return hardware_mapped_circuit, final_mapping
+
+    def transpiler_step(self, qibo_circuit):
+        """Transpilation step. Find new mapping, add swap gates and apply gates that can be run with this configuration.
+
+        Args:
+            qibo_circuit (:class:`qibo.models.Circuit`): circuit to be transpiled.
+        """
+        len_2q_circuit = len(self._circuit_repr)
+        path, meeting_point = self.relocate()
+        self.add_swaps(path, meeting_point)
+        self.update_qubit_map()
+        self.add_gates(qibo_circuit, len_2q_circuit - len(self._circuit_repr))
+
+    def first_transpiler_step(self, qibo_circuit):
+        """First transpilation step. Apply gates that can be run with the initial qubit mapping.
+
+        Args:
+            qibo_circuit (:class:`qibo.models.Circuit`): circuit to be transpiled.
+        """
+        self._circuit_position = 0
+        self.added_swaps = 0
+        len_2q_circuit = len(self._circuit_repr)
+        self._circuit_repr = self.reduce(self._graph)
+        self.add_gates(qibo_circuit, len_2q_circuit - len(self._circuit_repr))
+
+    @property
+    def sampling_split(self):
+        return self._sampling_split
+
+    @sampling_split.setter
+    def sampling_split(self, sampling_split):
+        """Set the sampling split.
+
+        Args:
+            sampling_split (float): define fraction of shortest path tested.
+        """
+
+        if sampling_split > 0.0 and 1.0 >= sampling_split:
+            self._sampling_split = sampling_split
+        else:
+            raise_error(ValueError, "Sampling_split must be in (0:1]")
+
+    def draw_connectivity(self):  # pragma: no cover
+        """Draw connectivity graph."""
+        pos = nx.spectral_layout(self.connectivity)
+        nx.draw(self.connectivity, pos=pos, with_labels=True)
+        plt.show()
+
+    def reduce(self, graph):
+        """Reduce the circuit, delete a 2-qubit gate if it can be applied on the current configuration.
+
+        Args:
+            graph (networkx.Graph): current hardware qubit mapping.
+
+        Returns:
+            new_circuit (list): reduced circuit.
+        """
+        new_circuit = self._circuit_repr.copy()
+        while new_circuit != [] and (new_circuit[0][0], new_circuit[0][1]) in graph.edges():
+            del new_circuit[0]
+        return new_circuit
+
+    def map_list(self, path):
+        """Return all possible walks of qubits, or a fraction, for a given path.
+
+        Args:
+            path (list): path to move qubits.
+
+        Returns:
+            mapping_list (list): all possible walks of qubits, or a fraction of them based on self.sampling_split, for a given path.
+            meeting_point_list (list): qubit meeting point for each path.
+        """
+        path_ends = [path[0], path[-1]]
+        path_middle = path[1:-1]
+        mapping_list = []
+        meeting_point_list = []
+        test_paths = list(range(len(path) - 1))
```

Comment:
```suggestion
        test_paths = range(len(path) - 1)
```
I don't think you need a list